packages/EDA/src/pls/pls-tools.ts

```diff
@@ -0,0 +1,389 @@
+    valueColumnName: result.regressionCoefficients.name,
+    valueAggrType: DG.AGG.AVG,
+    help: LINK.COEFFS,
+    showValueSelector: false,
+    showStackSelector: false,
+  }));
+
+  // 3. Loadings Scatter Plot
+  result.xLoadings.forEach((col, idx) => col.name = `${TITLE.XLOADING}${idx + 1}`);
+  const loadingsScatter = view.addViewer(DG.Viewer.scatterPlot(buffer, {
+    title: TITLE.LOADINGS,
+    xColumnName: `${TITLE.XLOADING}1`,
+    yColumnName: `${TITLE.XLOADING}${result.xLoadings.length > 1 ? '2' : '1'}`,
+    markerType: DG.MARKER_TYPE.CIRCLE,
+    labels: TITLE.FEATURE,
+    help: LINK.LOADINGS,
+  }));
+
+  // 4. Scores Scatter Plot
+
+  // 4.1) data
+  const scoreNames = [] as string[];
+
+  plsCols.forEach((col, idx) => {
+    col.name = cols.getUnusedName(`${TITLE.XSCORE}${idx + 1}`);
+    scoreNames.push(col.name);
+  });
+  result.uScores.forEach((col, idx) => {
+    col.name = cols.getUnusedName(`${TITLE.YSCORE}${idx + 1}`);
+    cols.add(col);
+    scoreNames.push(col.name);
+  });
+
+  // 4.2) create scatter
+  const scoresScatter = DG.Viewer.scatterPlot(input.table, {
+    title: TITLE.SCORES,
+    xColumnName: plsCols[0].name,
+    yColumnName: (plsCols.length > 1) ? plsCols[1].name : result.uScores[0],
+    markerType: DG.MARKER_TYPE.CIRCLE,
+    labels: input.names?.name,
+    help: LINK.SCORES,
+    showViewerFormulaLines: true,
+  });
+
+  // 4.3) create lines & circles
+  const lines = [] as DG.FormulaLine[];
+
+  scoreNames.forEach((xName) => {
+    const x = '${' + xName + '}';
+    lines.push({type: 'line', formula: `${x} = 0`, width: LINE_WIDTH, visible: true, title: ' ', color: COLOR.AXIS});
+
+    scoreNames.forEach((yName) => {
+      const y = '${' + yName + '}';
+
+      RADIUS.forEach((r) => {
+        const formula1 = y + ` = sqrt(${r*r} - ${x} * ${x})`;
+        lines.push({
+          type: 'line',
+          formula: formula1,
+          width: LINE_WIDTH,
+          visible: true,
+          title: ' ',
+          min: -r,
+          max: r,
+          color: COLOR.CIRCLE,
+        });
+
+        const formula2 = y + ` = -sqrt(${r*r} - ${x} * ${x})`;
+        lines.push({
+          type: 'line',
+          formula: formula2,
+          width: LINE_WIDTH,
+          visible: true,
+          title: ' ',
+          min: -r,
+          max: r,
+          color: COLOR.CIRCLE,
+        });
+      });
+    });
+  });
+
+  scoresScatter.meta.formulaLines.addAll(lines);
+  view.addViewer(scoresScatter);
+
+  // 5. Explained Variances
+
+  // 5.1) computation, source: the paper https://doi.org/10.1002/cem.2589
+  //      here, we use notations from this paper
+  const q = result.yLoadings.getRawData();
+  const p = result.xLoadings.map((col) => col.getRawData());
+  const n = input.table.rowCount;
+  const m = featuresNames.length;
+  const A = input.components;
+  const yExplVars = new Float32Array(A);
+  const compNames = [] as string[];
+  const xExplVars = [] as Float32Array[];
```

Comment:
I just wonder if this notation would be slightly better? Probably no difference, but here you specify the type explicitly vs inferring it. I don't have a strong opinion on that, but decided to mention the other way :)

const xExplVars: Float32Array[] = [];

Reply:
Modified! Your suggestion is more preferable.